fix: media files not being displayed

comatory opened the pull request:
Addresses https://github.com/eramdam/mastodon-archive-viewer/issues/1

I'm not sure why this did not work for me, I followed the original instructions in README but perhaps they changed the export format.

My problem was that images started with folder named after the instance:

```
/<instance_name>/media_attachments/files/xxx/xxx/xxx.jpeg
```

So my solution:

1. Removes the first part of the path to account for that
2. Adds the helper to `<a>` wrapping the media so clickin on it opens the correct URL

Commit:
fix: media files not being displayed

I'm not sure why this did not work for me, I followed the original
instructions in README but perhaps they changed the export format.

My problem was that images started with folder named after the instance:

```
/<instance_name>/media_attachments/files/xxx/xxx/xxx.jpeg
```

So my solution:

1. Removes the first part of the path to account for that
2. Adds the helper to `<a>` wrapping the media so clickin on it opens
   the correct URL

eramdam commented on .gitignore:
```diff
@@ -21,3 +21,7 @@ pnpm-debug.log*
+
+# do not version provided and build data
+archive-data/
+public/
```
that's a good catch, I should have added that prior.

comatory replied:
It would be perhaps better to remove existing files from that folder which are already checked in (with `git rm`), better approach would be to have a sample data in `docs/example` folder which you could copy to test it with. But I can do that in follow up PR.

comatory replied:
> better approach would be to have a sample data in docs/example folder which you could copy to test it with. But I can do that in follow up PR.

@eramdam I went ahead and created another [PR](https://github.com/eramdam/mastodon-archive-viewer/pull/3) just so you can see how I envisioned it. I think it makes stuff a bit more clearer and will make people to try it out with "Quick start" section.

eramdam reviewed (approved):
Looks good to me, thanks for catching this!

I'll do a quick test run against an export from my own instance (running Mastodon 4.3.1) just to make sure it still works before merging.

comatory commented:
@eramdam I noticed in your example data, your links look different. Can you try exporting your archive now? Maybe they changed it, if not, I'll investigate.

eramdam commented:
> @eramdam I noticed in your example data, your links look different. Can you try exporting your archive now? Maybe they changed it, if not, I'll investigate.

Here are attachments from an export I just ran using Mastodon 4.3.1

```json
        "attachment": [
          {
            "type": "Document",
            "mediaType": "image/png",
            "url": "/media_attachments/files/113/060/493/549/872/113/original/68528d6cfb0dd055.png",
            "name": null,
            "blurhash": "UQCPLaWC4=j?-nWCj[oK0gWV?Ff84=oet5R+",
            "width": 1280,
            "height": 1280
          },
          ...
        ],

// ...

 {
            "type": "Document",
            "mediaType": "video/mp4",
            "url": "/media_attachments/files/113/060/503/090/015/809/original/433c94e71bdf96ea.mp4",
            "name": null,
            "blurhash": "U11.~l%#Dg8^e-aekCkDI8M_t9x]Rja#t7oe",
            "width": 480,
            "height
}

// ...

  {
            "type": "Document",
            "mediaType": "audio/mpeg",
            "url": "/media_attachments/files/113/060/505/561/007/145/original/32a7be64599a4fdb.mp3",
            "name": null,
            "blurhash": null
          }
```

comatory commented:
@eramdam OK, maybe the instance I was on had different version. I don't know how to check it, I'm deactivated on that instance but I can still reach the old admin. I'll ask him/her. I tried looking into changelog and they don't mention anything about changes to archive, perhaps it's some server rule, depending on how the admin sets up media uploads.

comatory commented:
So I went ahead and downloaded archive from my current instance, I don't self host it (it's on masto.host) and they claim they run the latest stable release. It's still adding the additional folder containing the domain name 🤔 

```
        "attachment": [
          {
            "type": "Document",
            "mediaType": "video/mp4",
            "url": "/<mydomainnamecom>/media_attachments/files/113/549/099/427/972/470/original/video.mp4"
            "name": null,
            "blurhash": "xxxx",
            "width": 500,
            "height": 250
          }
        ],

```

eramdam commented:
> @eramdam OK, maybe the instance I was on had different version. I don't know how to check it

The version of Mastodon is usually displayed in the footer of the web UI, for example for mine 
![CleanShot 2024-11-29 at 13 34 51@2x](https://github.com/user-attachments/assets/f4396526-061f-42d0-a6c0-1545363c3c86)

But that's only part of the answer because if you generated your archive a while back, it might have been upgraded since then 😅  

> I tried looking into changelog and they don't mention anything about changes to archive, perhaps it's some server rule, depending on how the admin sets up media uploads.

Yeah I'll admit I do not know how that part works exactly. 
The `outbox.json` file is just a big array with all the ActivityPub objects that Mastodon sent over the lifetime of your account so I guess our best bet is to look at how Mastodon is (supposed to) create these. The examples in the [Mastodon documentation](https://docs.joinmastodon.org/spec/activitypub/) and the [Mastodon test suite](https://github.com/mastodon/mastodon/blob/2765cb52d298672778a73eafefba1fe80cf6175d/spec/lib/activitypub/activity/create_spec.rb#L560-L579) have absolute links so I don't even know if that's really helpful.

comatory commented:
Ok my current instance is also `4.3.1` 🫠 

<img width="332" alt="Screenshot 2024-11-29 at 22 42 24" src="https://github.com/user-attachments/assets/45a23889-d515-4915-9e3c-422beda9e402">

So this is probably some settings somewhere. I will augment this PR with some logic for rewrites.

eramdam commented:
> So I went ahead and downloaded archive from my current instance, I don't self host it (it's on masto.host) and they claim they run the latest stable release. It's still adding the additional folder containing the domain name 🤔
> 
> ```
>         "attachment": [
>           {
>             "type": "Document",
>             "mediaType": "video/mp4",
>             "url": "/<mydomainnamecom>/media_attachments/files/113/549/099/427/972/470/original/video.mp4"
>             "name": null,
>             "blurhash": "xxxx",
>             "width": 500,
>             "height": 250
>           }
>         ],
> ```

If I had to guess, this is probably the result of Masto.host running multiple instances on a single server/machine so the domain name being in the path doesn't surprise me too much.

Commit:
revert conversion of attachment URL during runtime

It seems like this situation, where attachment URL is prefixed
with server domain folder is not that common. So instead, this will
be a build time solution.

Added FAQ in case users run into this problem.

Commit:
type-check the script

Commit:
fix formatting

comatory commented:
Ok I pushed another commit. I chose a different approach and I will consider my situation to be an edge case. There's now FAQ section in the README pointing to this conversation. I added a build-time script that you can apply to normalize the paths, so one can optionally apply it if needed. I reverted the changes to data helper function.

I hope it's ok to use Javascript for the script, I did not want to deal with compiling it to TS. I added JSDoc annotation and have it type-check as well so I think all is good.

Let me know what you think (also about the other PR).

eramdam commented:
I'll have to take a close look at the PR to go forward, I haven't touched the project seriously in a white.
Can you, however, share the general folder/file structure of your archive (something like a truncated output of `tree` is enough)? 
I guess the crux of the issue comes from the fact that the files in the archive are under `archive.zip/media_attachments` and the URLs are `/yourinstancedotcom/media_attachments`, correct?[^1]

If so, I would, maybe, interpret this as a Mastodon bug? Not that we shouldn't work around it, but it's worth thinking about. 

[^1]: Perhaps this is what you meant all along, it's a holiday weekend here, so I'm not working at full speed 😂

eramdam commented on scripts/fix-attachments-url.mjs:
```diff
@@ -0,0 +1,101 @@
+  console.info(`Normalized ${changeCounter} attachment URLs`);
+}
+
+void main();
```
I wonder if, instead of having this script, it'd be better to introduce a method like `getMediaPath` inside the front-end that does what this script does? 

It should be fairly cheap to just transform the path by splitting over `media_attachments` to ensure it starts with `/media_attachments` and more importantly, IMO, doesn't require user-action or any "tampering" of the archive export.

what do you think?

comatory replied:
Yep I went ahead and did that. Seems like more straightforward solution with less code... but yeah it could be bit slower on big archives (that's why I was thinking of doing it in build time); but let's worry about that if it really happens.

See my last commits (I reverted the introduced change & then re-introduced slightly refactored code).

Commit:
Revert "fix formatting"

This reverts commit 99526b64429d8afc0264833550a6b47b1836eff9.

Commit:
Revert "type-check the script"

This reverts commit dd83eb0a7afccf28f5f7ac5451c833a69ee9ee7c.

Commit:
Revert "revert conversion of attachment URL during runtime"

This reverts commit 21839ef5960ba706c9f72356f53f72bd96dd54d3.

Commit:
handle non-standard media file paths in runtime

comatory commented:
> I'll have to take a close look at the PR to go forward, I haven't touched the project seriously in a white.

Sure - I re-did it in commit `57b296e0e1274d7179999a9434b5691831455c56`

> Can you, however, share the general folder/file structure of your archive (something like a truncated output of `tree` is enough)? I guess the crux of the issue comes from the fact that the files in the archive are under `archive.zip/media_attachments` and the URLs are `/yourinstancedotcom/media_attachments`, correct?[1](#user-content-fn-1-0046d37bf5c4223eb7a4ea67eb168660)

Sure, this is partial content of unzipped file (`~/Downloads/archive-20241126092459-d7bc1ebf89e491597fe4c372858052db
/Users/<user>/Downloads/archive-20241126092459-<uuid>/`):

```tree
├── actor.json
├── avatar.jpg
├── bookmarks.json
├── header.jpg
├── likes.json
├── media_attachments
│   └── files
│       ├── 000
│       │   ├── 014
│       │   │   ├── 167
│       │   │   │   └── original
│       │   │   │       └── 043797b2b6721f5f.jpg
│       │   │   ├── 317
│       │   │   │   └── original
│       │   │   │       └── bbdd92538bba06ee.jpg
│       │   │   ├── 328
│       │   │   │   └── original
│       │   │   │       └── 63021eae752ca377.mp4
│       │   │   ├── 583
│       │   │   │   └── original
│       │   │   │       └── 66321d96a2093008.mp4
│       │   │   ├── 587
│       │   │   │   └── original
│       │   │   │       └── 5af85d4ec1ae9383.jpg
│       │   │   ├── 717
│       │   │   │   └── original
│       │   │   │       └── 09872d1d1fcf0821.jpg
│       │   │   ├── 772
│       │   │   │   └── original
│       │   │   │       └── a5bd74b2c29f8202.jpg
│       │   │   └── 773
│       │   │       └── original
│       │   │           └── 7a33e3e520a5b476.jpg
│       │   ├── 015
│       │   │   ├── 031
│       │   │   │   └── original
│       │   │   │       └── 55a1ba5708b7b404.jpg
│       │   │   ├── 451
```

So the folder structure seems to be normal but `url` in attachments seems to point somewhere else.

> If so, I would, maybe, interpret this as a Mastodon bug? 

I lean towards what you mentioned before - it's probably a hosting thing on masto.host
